### PR TITLE
Fix ExtractInterceptor for subclasses

### LINKED ISSUE #2508: OnPreUpdateCollection - Passed entity instance X is not of expected type Y
Hi!

When updating from version 5.2.7 to 5.3.2, we get the following error in OnPreUpdateCollection event listener.

Seems to occur when changing a collection in a class (joined subclass) with a lazy property .

Using the lazy property gives us the ProxyForFieldInterceptor and the exception.

at NHibernate.Tuple.Entity.BytecodeEnhancementMetadataPocoImpl.ExtractInterceptor(Object entity)
   at NHibernate.Tuple.Entity.BytecodeEnhancementMetadataPocoImpl.HasAnyUninitializedLazyProperties(Object entity)
   at NHibernate.Tuple.Entity.AbstractEntityTuplizer.ShouldGetAllProperties(Object entity)
   at NHibernate.Tuple.Entity.PocoEntityTuplizer.GetPropertyValues(Object entity)
   at NHibernate.Persister.Entity.AbstractEntityPersister.GetPropertyValues(Object obj)
   at NHibernate.Test.NHSpecificTest.GH2508.AuditEventListener.OnPreUpdateCollection(PreCollectionUpdateEvent event) in 

Thank you for keeping NH alive!

### PR DESCRIPTION
**Pull request by maca88**
Fix for a regression made by #1946.

Fixes #2508

### CODE REVIEW

File: src/NHibernate.Test/NHSpecificTest/GH2508/Mappings.hbm.xml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (bahusoid) [inline src/NHibernate.Test/NHSpecificTest/GH2508/Mappings.hbm.xml:41]:
```suggestion
```
Don't include me in co-authors for this priceless contribution :)

**Review by bahusoid — APPROVED** [review]

**Comment by bahusoid** [comment]
What about check inside `InjectInterceptor`:
https://github.com/nhibernate/nhibernate-core/blob/9fc784e6a3dac2df3518ca612e461d75d3aca7dc/src/NHibernate/Tuple/Entity/BytecodeEnhancementMetadataPocoImpl.cs#L135

Shouldn't it be `_entityType.IsInstanceOf(entity)`

**Comment by maca88** [comment]
> Shouldn't it be _entityType.IsInstanceOf(entity)?

Good observation, yes it should. Changed.

File: src/NHibernate/Tuple/Entity/BytecodeEnhancementMetadataPocoImpl.cs [hunk 2]
```diff
@@ -132,7 +132,7 @@ public IFieldInterceptor InjectInterceptor(object entity, ISessionImplementor se
 				return null; // Can happen when a saved entity is refreshed within the same session NH2860
 			}
 
-			if (entity.GetType().BaseType != _entityType)
+			if (_entityType.IsInstanceOf(entity))
```
Comment (bahusoid) [inline src/NHibernate/Tuple/Entity/BytecodeEnhancementMetadataPocoImpl.cs:135]:
```suggestion
			if (_entityType.IsInstanceOfType(entity))
```

File: src/NHibernate/Tuple/Entity/BytecodeEnhancementMetadataPocoImpl.cs [hunk 3]
```diff
@@ -132,7 +132,7 @@ public IFieldInterceptor InjectInterceptor(object entity, ISessionImplementor se
 				return null; // Can happen when a saved entity is refreshed within the same session NH2860
 			}
 
-			if (entity.GetType().BaseType != _entityType)
+			if (_entityType.IsInstanceOfType(entity))
```
Comment (hazzik) [inline src/NHibernate/Tuple/Entity/BytecodeEnhancementMetadataPocoImpl.cs:135]: Shouldn't it be `!_entityType.IsInstanceOfType(entity)`?

**Review by hazzik — APPROVED** [review]

**Pull request merged**